File: 20 - Find Longest Substring Without Repeating Characters/solution.js
function findLongestSubstring(string) {
  // Your code here
  let maxLength = 0;
  let start = 0;
  const seen = new Map();

  for (let end = 0; end < string.length; end++) {
    const char = string[end];

    if (seen.has(char) && seen.get(char) >= start) {
      start = seen.get(char) + 1;
    }

    seen.set(char, end);

    maxLength = Math.max(maxLength, end - start + 1);
  }

  return maxLength;
}

//   Test cases
console.log(findLongestSubstring("abcabcbb"));

File: 20 - Find Longest Substring Without Repeating Characters/solution.py
def find_longest_substring(string):
    # Your code here
    max_length = 0
    start = 0
    seen = dict()

    for end in range(len(string)):
        char = string[end]

        if char in seen and seen[char] >= start:
            start = seen[char] + 1

        seen[char] = end

        max_length = max(max_length, end - start + 1)

    return max_length


#   Test cases
print(find_longest_substring("abcabcbb"))